File: qabelbox/src/main/java/de/qabel/qabelbox/storage/notifications/AndroidStorageNotificationPresenter.java
package de.qabel.qabelbox.storage.notifications;


import android.app.Notification;
import android.content.Context;
import android.content.Intent;
import android.support.v7.app.NotificationCompat;

import org.apache.commons.io.FileUtils;

import de.qabel.qabelbox.R;
import de.qabel.qabelbox.base.ActiveIdentityActivity;
import de.qabel.qabelbox.base.MainActivity;
import de.qabel.qabelbox.notifications.QblNotificationPresenter;

public class AndroidStorageNotificationPresenter extends QblNotificationPresenter
        implements StorageNotificationPresenter {

    private static final String TAG = AndroidStorageNotificationPresenter.class.getSimpleName();

    public AndroidStorageNotificationPresenter(Context context) {
        super(context);
    }

    @Override
    protected String getTag() {
        return TAG;
    }

    public Intent createFileIntent(StorageNotificationInfo info) {
        Intent notificationIntent = new Intent(getContext(), MainActivity.class);
        notificationIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
                | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        notificationIntent.putExtra(ActiveIdentityActivity.Constants.ACTIVE_IDENTITY, info.getIdentityKeyId());
        notificationIntent.putExtra(MainActivity.START_FILES_FRAGMENT, true);
        notificationIntent.putExtra(MainActivity.START_FILES_FRAGMENT_PATH, info.getPath());
        return notificationIntent;
    }

    @Override
    public void updateUploadNotification(int queueSize, StorageNotificationInfo info) {
        int notificationId = getIdForInfo(info);
        Intent fileIntent = createFileIntent(info);
        NotificationCompat.Builder notificationBuilder;
        if (!info.isComplete()) {
            String title = getContext().getResources().
                    getQuantityString(R.plurals.uploadsNotificationTitle,
                            queueSize, (queueSize > 1 ? queueSize : info.getFileName()));
            String content = String.format(
                    getString(R.string.upload_in_progress_notification_content),
                    info.getProgress() + "%");
            notificationBuilder = createNotification(
                    fileIntent, title, R.drawable.cloud_upload, content);
            notificationBuilder.setProgress(100, info.getProgress(), false);
        } else {
            notificationBuilder = createNotification(
                    fileIntent,
                    String.format(getString(R.string.upload_complete_notification_title)),
                    R.drawable.cloud_upload,
                    String.format(getString(R.string.upload_complete_notification_msg),
                            info.getFileName()));
        }
        showNotification(notificationId, notificationBuilder);
    }

    private void showNotification(int id, NotificationCompat.Builder builder) {
        notify(id, builder, Notification.CATEGORY_PROGRESS, false);
    }

    @Override
    public void updateDownloadNotification(StorageNotificationInfo info) {
        Intent fileIntent = createFileIntent(info);
        String title, msg;
        if (!info.isComplete()) {
            title = String.format(getString(R.string.downloading), info.getFileName());
            msg = FileUtils.byteCountToDisplaySize(info.getDoneBytes())
                    + " / " + FileUtils.byteCountToDisplaySize(info.getTotalBytes());
        } else {
            title = getString(R.string.download_complete);
            msg = String.format(getString(R.string.download_complete_msg), info.getFileName());
        }
        NotificationCompat.Builder notificationBuilder =
                createNotification(fileIntent, title, R.drawable.download, msg);

        if (!info.isComplete()) {
            notificationBuilder.setProgress(100, info.getProgress(), false);
        }
        showNotification(getIdForInfo(info), notificationBuilder);
    }
}

File: qabelbox/src/main/java/de/qabel/qabelbox/box/notifications/AndroidStorageNotificationPresenter.kt
package de.qabel.qabelbox.box.notifications


import android.app.Notification
import android.content.Context
import android.content.Intent
import android.support.v7.app.NotificationCompat
import de.qabel.qabelbox.R
import de.qabel.qabelbox.base.MainActivity
import de.qabel.qabelbox.notifications.QblNotificationPresenter
import org.apache.commons.io.FileUtils

class AndroidStorageNotificationPresenter(context: Context) :
        QblNotificationPresenter<String, StorageNotificationInfo>(context), StorageNotificationPresenter {

    companion object {
        private val TAG = AndroidStorageNotificationPresenter::class.java.simpleName
        private const val UPLOAD_ICON = R.drawable.cloud_upload
        private const val DOWNLOAD_ICON = R.drawable.download
    }

    override fun getTag(): String = TAG

    fun createFileBrowserIntent(info: StorageNotificationInfo): Intent =
            Intent(context, MainActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                putExtra(MainActivity.ACTIVE_IDENTITY, info.identityKeyId)
                putExtra(MainActivity.START_FILES_FRAGMENT, true)
                putExtra(MainActivity.START_FILES_FRAGMENT_PATH, info.path)
            }

    private fun showNotification(info: StorageNotificationInfo, builder: NotificationCompat.Builder,
                                 autoCancel: Boolean = false) {
        builder.setWhen(info.time)
        builder.setAutoCancel(autoCancel)
        notify(getIdForInfo(info), builder, Notification.CATEGORY_PROGRESS, autoCancel)
    }

    private fun formatProgress(info: StorageNotificationInfo): String =
            "${FileUtils.byteCountToDisplaySize(info.doneBytes)} / " +
                    "${FileUtils.byteCountToDisplaySize(info.totalBytes)}\t ${info.progress}%"

    override fun showEncryptingUploadNotification(info: StorageNotificationInfo) {
        createNotification(createFileBrowserIntent(info),
                getString(R.string.uploading, info.fileName), UPLOAD_ICON,
                getString(R.string.encrypting)).let {
            it.setProgress(10, 10, true)
            showNotification(info, it, false)
        }
    }

    override fun showUploadProgressNotification(info: StorageNotificationInfo) {
        //TODO Cancel action
        createNotification(createFileBrowserIntent(info),
                getString(R.string.uploading, info.fileName), UPLOAD_ICON,
                formatProgress(info)).let {
            it.setProgress(100, info.progress, false)
            showNotification(info, it, false)
        }
    }

    override fun showUploadCompletedNotification(info: StorageNotificationInfo) {
        createNotification(createFileBrowserIntent(info),
                getString(R.string.upload_complete_title),
                UPLOAD_ICON,
                getString(R.string.upload_complete_msg, info.fileName)).let {
            showNotification(info, it, true)
        }
    }

    override fun showUploadFailedNotification(info: StorageNotificationInfo) {
        //TODO Retry action
        createNotification(createFileBrowserIntent(info),
                getString(R.string.upload_failed_title),
                UPLOAD_ICON,
                getString(R.string.upload_failed_msg, info.fileName)).let {
            showNotification(info, it, true)
        }
    }

    override fun showDownloadProgressNotification(info: StorageNotificationInfo) {
        //TODO Cancel action
        createNotification(createFileBrowserIntent(info),
                getString(R.string.downloading, info.fileName), DOWNLOAD_ICON,
                formatProgress(info)).let {
            it.setProgress(100, info.progress, false)
            showNotification(info, it, false)
        }
    }

    override fun showDecryptingDownloadNotification(info: StorageNotificationInfo) {
        createNotification(createFileBrowserIntent(info),
                getString(R.string.downloading, info.fileName), DOWNLOAD_ICON,
                getString(R.string.decrypting)).let {
            it.setProgress(10, 10, true)
            showNotification(info, it, false)
        }
    }

    override fun showDownloadCompletedNotification(info: StorageNotificationInfo) {
        createNotification(createFileBrowserIntent(info),
                getString(R.string.download_complete),
                UPLOAD_ICON,
                getString(R.string.download_complete_msg, info.fileName)).let {
            showNotification(info, it, true)
        }
    }

    override fun showDownloadFailedNotification(info: StorageNotificationInfo) {
        //TODO Retry action
        createNotification(createFileBrowserIntent(info),
                getString(R.string.download_failed), DOWNLOAD_ICON,
                getString(R.string.download_failed_msg, info.fileName)).let {
            showNotification(info, it, true)
        }
    }
}
